### PR TITLE
Handle error

### PR DESCRIPTION
**Pull request by sebing**
Fixing issue
https://github.com/yegor256/takes/issues/522

### CODE REVIEW

**Comment by davvd** [comment]
@sebing Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@pinaf it's yours,please review

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 1]
```diff
@@ -153,7 +153,7 @@ private String token(final String home, final String code)
         final String uri = new Href(this.github)
             .path("login").path("oauth").path("access_token")
             .toString();
-        return new JdkRequest(uri)
+        final XmlResponse xmlres = new JdkRequest(uri)
```
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing let's name this variable `response`.

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 2]
```diff
@@ -164,8 +164,16 @@ private String token(final String home, final String code)
             .back()
             .fetch().as(RestResponse.class)
             .assertStatus(HttpURLConnection.HTTP_OK)
-            .as(XmlResponse.class)
-            .xml().xpath("/OAuth/access_token/text()").get(0);
+            .as(XmlResponse.class);
+        String tokenString = null;
+        try {
+            tokenString = xmlres.xml()
```
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing let's move the `return` statement here.

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 3]
```diff
@@ -164,8 +164,16 @@ private String token(final String home, final String code)
             .back()
             .fetch().as(RestResponse.class)
             .assertStatus(HttpURLConnection.HTTP_OK)
-            .as(XmlResponse.class)
-            .xml().xpath("/OAuth/access_token/text()").get(0);
+            .as(XmlResponse.class);
+        String tokenString = null;
```
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing and get rid of this variable.

File: src/main/java/org/takes/facets/auth/social/PsGithub.java (hunk 1 above)
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing let's keep the return here and move the `try` to the line above.

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 4]
```diff
@@ -180,7 +188,7 @@ private static Identity parse(final JsonObject json) {
         props.put("login", json.getString("login", "unknown"));
         props.put("avatar", json.getString("avatar_url", "#"));
         return new Identity.Simple(
-            String.format("urn:github:%d", json.getInt("id")), props
```
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing please revert this change

**Comment by pinaf** [comment]
@sebing we need a unit test for the new behavior.

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 5]
```diff
@@ -164,8 +164,16 @@ private String token(final String home, final String code)
...
-            .as(XmlResponse.class)
-            .xml().xpath("/OAuth/access_token/text()").get(0);
+            .as(XmlResponse.class);
+        String tokenString = null;
+        try {
+            tokenString = xmlres.xml()
+                .xpath("/OAuth/access_token/text()").get(0);
+        } catch (final IndexOutOfBoundsException exce) {
+            throw new HttpException(
+                HttpURLConnection.HTTP_BAD_REQUEST,"Invalid XML",exce);
```
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing please add whitespaces after each `,`

Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing also, let's move the ending `);` to the next line, aligned with the `throw`.

**Comment by sebing** [comment]
@pinaf thanks for the suggestions, I corrected them. 
I created the unit test case for the changes too. 

Please verify and let me know your views.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 6]
```diff
@@ -148,4 +176,29 @@ private static void assertParam(final Request req,
             Matchers.equalTo(value)
         );
     }
+
+    /**
+     * An inner class for the Take implementation testing.
+     */
+    private static class PsGithubTestTake implements Take {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing why create an inner class if it's only used in 1 place?

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 7]
```diff
@@ -135,6 +152,17 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * A private method to create the basic directive.
+     * without the access token.
+     * @return A basic directive.
+     */
+    private static Directives getDirectiveWithoutAccessToken() {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing let's rename this to `basicDirectives`.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 8]
```diff
@@ -135,6 +152,17 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * A private method to create the basic directive.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing basic directive`s, without access token.`

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 9]
```diff
@@ -54,12 +57,51 @@
  */
 public final class PsGithubTest {
     /**
+     * A constant for GitHubToken.
+     */
+    private static final String GIT_HUB_TOKEN = "GitHubToken";
+
+    /**
+     * A constant for access_token.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing isn't this a header? If so, let's go with `Access token HTTP header.`

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 10]
```diff
@@ -54,12 +57,51 @@
  */
 public final class PsGithubTest {
     /**
+     * A constant for GitHubToken.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing `GitHubToken`

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 11]
```diff
@@ -54,12 +57,51 @@
...
+     */
+    @Rule
+    public transient ExpectedException thrown = ExpectedException.none();
+
+    /**
+     * PsGithub can throw HttpExeption.
+     * @throws IOException If some problem inside.
+     */
+    @Test
+    public void throwHttpException() throws IOException {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing this needs a better name. perhaps `failsOnNoAccessToken`

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 12]
```diff
@@ -54,12 +57,51 @@
...
+     */
+    private static final String ACCESS_TOKEN = "access_token";
+    /**
+     * A Junit Exception test variable.
+     */
+    @Rule
+    public transient ExpectedException thrown = ExpectedException.none();
+
+    /**
+     * PsGithub can throw HttpExeption.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing PsGithub can `fail when no access token is present`.

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 13]
```diff
@@ -153,19 +153,25 @@ private String token(final String home, final String code)
...
+                .formParam("client_secret", this.key)
+                .formParam("code", code)
+                .back()
+                .fetch().as(RestResponse.class)
+                .assertStatus(HttpURLConnection.HTTP_OK)
+                .as(XmlResponse.class)
+                .xml().xpath("/OAuth/access_token/text()").get(0);
+        } catch (final IndexOutOfBoundsException exception) {
+            throw new HttpException(
+                HttpURLConnection.HTTP_BAD_REQUEST, "Invalid XML", exception
```
Comment (pinaf) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing let's change the error message here to `No access token` instead of the generic `Invalid XML`.

**Comment by pinaf** [comment]
@sebing a few more comments and we are good.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 6 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: Anonymous inner class max allowed is 20

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 7 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf  I thought it was very specific directive without the access token, If we rename it to basicDirectives, tomorrow people may confuse?

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 8 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: Corrected the mistake.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 11 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: Corrected.

**Comment by sebing** [comment]
@pinaf I corrected the suggestions. (except renaming the method to  basic Directive)

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 14]
```diff
@@ -148,4 +175,29 @@ private static void assertParam(final Request req,
             Matchers.equalTo(value)
         );
     }
+
+    /**
+     * An inner class for the Take implementation testing.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing in that case let's change this to something more detailed. How about
`A test Take that verifies if the request contains the access token and returns a JSON login response` ?

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 6 above)
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing would also be a good idea to rename the class to `TkFakeLogin` or something similar.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 7 above)
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing I see your point. Let's remove the `get` though. `directivesWithoutAccessToken` would be nice.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 15]
```diff
@@ -135,6 +152,16 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * A private method to create the basic directives, without access token.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing no need to mention `A private method to...`. `Creates basic directives without an access token.` is fine.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 16]
```diff
@@ -54,12 +57,51 @@
...
+     */
+    private static final String ACCESS_TOKEN = "access_token";
+    /**
+     * A Junit Exception test variable.
+     */
+    @Rule
+    public transient ExpectedException thrown = ExpectedException.none();
+
+    /**
+     * PsGithub can fails on no access token.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing can fail, without the `s`.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 9 above)
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing this one still needs change.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 17]
```diff
@@ -54,12 +57,51 @@
...
     public void canLogin() throws IOException {
+        this.performLogin(
+            PsGithubTest.getDirectiveWithoutAccessToken()
+                .add(PsGithubTest.ACCESS_TOKEN)
+                .set(PsGithubTest.GIT_HUB_TOKEN)
+        );
+    }
+
+    /**
+     * A private method to perform the basic login.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing `Performs a basic login` is fine.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 18]
```diff
@@ -54,12 +57,51 @@
...
+     * @param directive The directive object.
+     * @throws IOException If some problem inside.
+     * @checkstyle MultipleStringLiteralsCheck (100 lines)s
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing do we really need this? also, there is an extra `s` at the end.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 6 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf ( i will push the change with other changes)

Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing also, please add `final` to this class.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 9 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf Added the comment with XPath.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 15 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf Corrected.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 16 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf Corrected.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 17 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf Corrected.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 18 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf Removed the line. (this line was there in the code already).

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java (hunk 6 above)
Comment (sebing) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @pinaf done.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 19]
```diff
@@ -54,12 +57,50 @@
...
     /**
+     * GitHubToken.
+     */
+    private static final String GIT_HUB_TOKEN = "GitHubToken";
+
+    /**
+     * XPath element access_token.
+     */
+    private static final String ACCESS_TOKEN = "access_token";
+    /**
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing let's add a blank line here for consistency.

File: src/test/java/org/takes/facets/auth/social/PsGithubTest.java [hunk 20]
```diff
@@ -135,6 +151,16 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * Creates the basic directives, without access token.
+     * @return A basic directive.
+     */
+    private static Directives getDirectiveWithoutAccessToken() {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/auth/social/PsGithubTest.java]: @sebing Let's remove the get. `directivesWithoutAccessToken` would be nice.

**Comment by sebing** [comment]
@pinaf 

> The javadoc is can fail but the method is fails.

I thought there should be consistancy.

**Comment by pinaf** [comment]
@sebing there is - it's just english. the test "fails", but the test "can fail" :)

**Comment by sebing** [comment]
@pinaf I corrected the suggestions .

**Comment by pinaf** [comment]
@rultor merge

**Comment by pinaf** [comment]
@sebing good job! thanks!

**Comment by sebing** [comment]
@pinaf  Thanks again for the suggestions and help.

**Comment by sebing** [comment]
@yegor256 @davvd Any update on the closure of this merge?

File: src/main/java/org/takes/facets/auth/social/PsGithub.java [hunk 21]
```diff
@@ -153,19 +153,25 @@ private String token(final String home, final String code)
...
+                .formParam("client_id", this.app)
+                .formParam("redirect_uri", home)
+                .formParam("client_secret", this.key)
+                .formParam("code", code)
+                .back()
+                .fetch().as(RestResponse.class)
+                .assertStatus(HttpURLConnection.HTTP_OK)
+                .as(XmlResponse.class)
+                .xml().xpath("/OAuth/access_token/text()").get(0);
+        } catch (final IndexOutOfBoundsException exception) {
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing this looks ugly. should be like this instead:

```
List<String> tokens = xpath();
if (tokens.isEmpty()) {
  throw...
}
return tokens.get(0);
```

**Comment by yegor256** [comment]
@sebing see my comments above

File: src/main/java/org/takes/facets/auth/social/PsGithub.java (hunk 21 above)
Comment (sebing) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @yegor256 The problem is the call to xpath is throwing the exception, so it won't reach the tokens.isEmpty line.

- I would like to have the return statement in the last line of the method, but @pinaf suggested this is the better way

Comment (yegor256) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing what exception you're talking about?

Comment (sebing) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @yegor256 The actual problem, we are trying to fix.
```
com.jcabi.xml.ListWrapper$NodeNotFoundException: XPath '/OAuth/access_token/text()' not found in
```
Is thrown in the call the `XML.xpath()` method.

Comment (yegor256) [inline src/main/java/org/takes/facets/auth/social/PsGithub.java]: @sebing it's thrown only if you call `get()`, but not `isEmpty()`

**Comment by sebing** [comment]
@yegor256 I refactored the code. Can you have a look at them, and accept them?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@ypshenychka please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by ypshenychka** [comment]
@sebing 
According to our [QA Rules](http://at.teamed.io/qa.html):
>Pull request title explains the problem it is fixing.
Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

Please correct title and description according to above rules.

**Comment by ypshenychka** [comment]
@sebing 
According to our [QA Rules](http://at.teamed.io/qa.html):
>Messages in a ticket always start with a name of a user they are addressed to.

Please correct couple of your code review comments by indicating an addressee in the beginning.

**Comment by sebing** [comment]
@ypshenychka 
- updated the PR title.
- updated the PR description.
- added missing addressee in the comments. 

Thanks for the suggestions. I will try to be more careful next time :)

**Comment by ypshenychka** [comment]
@sebing One more left :)
>sebing added a note 5 days ago
Corrected with a new name.

**Comment by sebing** [comment]
@ypshenychka sorry for missing out one . I corrected them.

**Comment by ypshenychka** [comment]
@sebing Thank you.

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
@pinaf 10 mins was added to the account of @ypshenychka (for QA review), in transaction 76577111; I added **44 mins** to your account, many thanks for working with the project (AP-45Y90720W83226110)! The completion time here was 142 hours and 50 mins.; extra minutes for review comments (c=29); added +44 to your rating, now it is equal to [+6862](http://www.netbout.com/b/35128?open=rating)
